### PR TITLE
Unit test for neuromodulation

### PR DESCRIPTION
**Pull request by WeinaJi**
## Context
This PR adds a circuit and unit tests for neuromodulation events. A brief introduction is also added in the docs.

## Scope
- `create.data.py`: create virtual nodes file and neuromodulatory projection edges file
- `neuromodulation_manager.py`: improve the code to find the closest post cell synapses in all connections including multiple populations.

## Testing
`test_simple_neuromodulation.py`

## Review
* [x] PR description is complete
* [x] Coding style (imports, function length, New functions, classes or files) are good
* [x] Unit/Scientific test added
* [x] Updated Readme, in-code, developer documentation

### CODE REVIEW

File: docs/source/neuromodulation.rst [hunk 1]
```diff
@@ -0,0 +1,22 @@
+===============
+Neuromodulation
+===============
+
+A neuromodulatory event influences synaptic behaviours by simulating additional neurontransmitters. It is implemented by linking from projection fibers to existing synapses as illustrated below:
+
+.. image:: ./img/neuromodulation.png
```
Comment (cattabiani) [inline docs/source/neuromodulation.rst:7]: it would be nice to have the picture in as neuromodulation.darwio.svg so it can be easily changed in the future

Comment (WeinaJi) [reply docs/source/neuromodulation.rst:7]: I didn't draw the diagram but captured from scientiest's slides. I tried to convert from png to drawio.svg , but vscode is not able to open it.

File: neurodamus/neuromodulation_manager.py [hunk 2]
```diff
@@ -74,7 +75,9 @@ def finalize(
             syn_params = self._synapse_params[syn_i]
             # We need to get all connections since we dont know the sgid
             # TODO: improve this by extracting all the relative distances only once
-            base_conns = base_manager.get_connections(self.tgid)
+            base_conns = chain(
+                *[base_manager.get_connections(self.tgid) for base_manager in base_managers]
+            )
```
Comment (cattabiani) [inline neurodamus/neuromodulation_manager.py:80]:
```suggestion
base_conns = chain.from_iterable(
    base_manager.get_connections(self.tgid) for base_manager in base_managers
)
```

Comment (cattabiani) [reply neurodamus/neuromodulation_manager.py:80]: ... if you want to flatten this is more efficient

File: tests/simulations/ringtest/create_data.py [hunk 3]
```diff
@@ -281,6 +286,42 @@ def make_ringtest_edges():
     }
     make_edges(filename="local_edges_C.h5", edges=edges, wanted_attributes=wanted_attributes)
 
+# For neuromodulation test: Create another A->B edge with different afferent_section_pos w.r.t B->B edge
```
Comment (cattabiani) [inline tests/simulations/ringtest/create_data.py:289]: if we do not pick the neuromodulation circuit this new file is unused right? In other words, this does not alter the behavior of the previous tests right?

File: tests/unit/test_simple_neuromodulation.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
+"""Test neuromodulations using the toy ringtest data as such:
+neuromodulatory projection: virtual_neurons->RingB, 0->0, 1->0
+The post synatic cell RingB gid 0 has 2 synapses:
+    RingA->RingB : 0->0
+    RingB->RingB : 1->0
+The neuromodulatory spikes are injected via replay from virtual neuron gid 1
```
Comment (cattabiani) [inline tests/unit/test_simple_neuromodulation.py:6]: is it gid 1 or 1001? otherwise... how can you apply neuromodulation to the synapse ringA (0) to ringB (1000)?

File: tests/unit/test_simple_neuromodulation.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+Therefore, the neuromodulator applies netcon to the synapse from connection RingB->RingB : 1->0
+"""
```
Comment (cattabiani) [inline tests/unit/test_simple_neuromodulation.py:8]: if I understand correctly it is neurodamus that assigns a certain synapse to a certain neuromodulation based on "proximity". We are not really testing that part here as we just assume that it gets the correct synapse and we roll with that. 

Maybe you could add a test to verify that this assignment works even for the edge cases

**Review by cattabiani — COMMENTED** [review]
Very nice PR! I am just concerned that we may under-test this part slightly. A few questions were added

File: tests/unit/test_simple_neuromodulation.py (hunk 4 above)
Comment (WeinaJi) [reply tests/unit/test_simple_neuromodulation.py:6]: Here in the doc string are all SONATA gid (without offset), in the tests I checked with the offset.

File: tests/unit/test_simple_neuromodulation.py (hunk 5 above)
Comment (WeinaJi) [reply tests/unit/test_simple_neuromodulation.py:8]: We don't keep that `found` synapse obj to the end, rather to store the `netcon` targeting that synapse if a replay is injected. Let me try what I can do more.

File: neurodamus/neuromodulation_manager.py (hunk 2 above)
Comment (WeinaJi) [reply neurodamus/neuromodulation_manager.py]: Good point, done!

File: tests/unit/test_simple_neuromodulation.py (hunk 5 above)
Comment (WeinaJi) [reply tests/unit/test_simple_neuromodulation.py:8]: A new test is added. Please check

**Review by cattabiani — APPROVED** [review]
LGTM
